### DISCUSSION TITLE
0040 XLS-40d: Decentralized Identity on XRPL

### DISCUSSION THREAD

**Discussion by aanchal4** [XRPLF/XRPL-Standards#100] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by ximinez** [comment]
A few thoughts:
1. The table for `ControllerAccount` under `DIDTokenCreate` is malformed and not displaying as a table.
2. The field `DIDToken.ControllerAccount` is marked as required. That seems redundant and wasteful of space for accounts which create their own `DIDToken`. A simple change would be to specify that if the `ControllerAccount` field is absent, treat the `Account` as controller.
4. The `ControllerAuthProof` can be provided and work with any of the transaction types. While the controllers need to have some degree of trust in the account submitting the transaction, a malicious submitter could turn, say, an update into a delete. They would probably only get one chance to do that, but it might be worth it for a high-value target. I suggest an `Operation` field be added to the `ControllerAuthProof` that species `create`, `update`, or `delete`.
5. Speaking of an `Operation` field, the three new transaction types are almost completely redundant (particularly `DIDTokenCreate` and `DIDTokenUpdate`. I'm picturing a lot of unnecessarily duplicated code in the implementation (or some worker functions). What if instead of three transactions, we specify one transaction, `DIDTokenSet` (similar to `AccountSet` and `TrustSet`) that performs the appropriate action based on `Operation`? This could simplify the implementation significantly. In fact, the three `Operations` could be reduced to two: `set`, and `delete`. `set` would create the object if it doesn't already exist, and update it if it does.

**Reply by aanchal4** [reply]
Thanks Ed!
1. That was supposed to be a comment. It did not render correctly. Will fix it.
2. That makes sense.
3. An update requires `URI` field whereas a delete MUST not have one. One can not swap delete for an update or vice-versa. Nor can any transaction be replayed. I don't see why we would need an Operation field for this reason.
4. We can combine Create and Update operations .

**Comment by politicaljabo** [comment]
Great work.

**Comment by ZhangxiangHu-Ripple** [comment]
Why not directly store the owner's public key and the controller's public key in the DIDToken object? In the current design, if I want to verify a user's DID, I need to call the resolver to get the URI first, then use the URI to retrieve the public key. This would introduce additional costs and delays.

**Reply by aanchal4** [reply]
The proposal recommends to add support for widely accepted standards based Decentralized Identifier (`DID`) creation and `CRUD` methods on XRPL. This proposal chooses [W3C DID](https://www.w3.org/TR/did-core/) and [DID document](https://www.w3.org/TR/did-spec-registries/) standards to be supported on XRPL. Note that following a standards-based approach is critical for:
- Interoperability: DID owner’s can use identity data (DID document and VCs) associated with their DIDs across services.), 
- Developers to use open-source tools such as [Universal Resolver](https://github.com/decentralized-identity/universal-resolver) that implement W3C standards for DID creation and CRUD operations for all blockchains that support those standards. Here is the list. We will add support for XRPL DID DID and CRUD operations to [DID-method registry](https://w3c.github.io/did-spec-registries/#did-methods) and to [Universal Resolver](https://github.com/decentralized-identity/universal-resolver) as well.

Per W3C DID standard, DID must resolve to a DID document. DID document is a standard format to store cryptographic keys, service end-points and other non-sensitive public information necessary to verify DID. It may also contain different keys for different services, etc. For details see [here](https://www.w3.org/TR/did-spec-registries/). In the proposal, there are two ways DID controller can choose to associate their DID with a DID document:
- Retrievable via a `URI` reference (useful for off-ledger services)
- Host it directly in the `Data` field (useful for ledger transactors)

The retrieved standard DID document can be used by services across (blockchain or non-blockchain) ecosystems that follow DID and DID document standards. Tools such as [Universal resolver](https://github.com/decentralized-identity/universal-resolver) that implement DIDs per W3C standards require a DID doc. Identity-based application developers can easily use these tools without having to implement DIDs for each blockchain. 

So it is possible to store a public key in the DIDToken object, but then the XRPL users and developers will lose all the benefits of a standards-based approach.

That is, a DID controller may choose not to follow DID documents and VC standards for data associated with DIDs (and it might work for specific use-cases), it is highly recommended to follow standards to reap the maximum benefits of this proposal.

**Reply by adrianhopebailie** [reply]
While this indirection seems inefficient I think it is the correct choice.

It definitely makes this design nice and clean.

The DID Document might define multiple keys for various purposes (e.g. authentication, signing etc) and it is likely that in most cases the entity that controls the account on XRPL (DID controller) shouldn't be using the same keys for signing transactions on the ledger.

**Comment by toshiyotmoto** [comment]
There are *many* Decentralized Identity patent applications. Since this proposal is the composure of years of work / knowledge, the **XRPL Foundation needs to submit any prior art, including this proposal, along with any notes, communications, exhibits, or any other form of publicly available information that existed prior to the filing date of any decentralized identity pending patent applications.**

Prior art serves to establish the state of existing knowledge and technology in a given field, which is crucial in determining whether an invention is novel and non-obvious, which is required in obtaining a patent. **In essence, a party cannot be granted a patent if their claims incorporate elements of domain knowledge (this proposal) that predate any patent or patent application.**

Here are a few patent applications the Foundation needs to consider submitting prior art for:

https://patents.google.com/patent/US20210272106A1
https://patents.google.com/patent/US20220215132A1
https://patents.google.com/patent/US20220067717A1
https://patents.google.com/patent/US20230050540A1
https://patents.google.com/patent/US20230046965A1
https://patents.google.com/patent/US20230059739A1
https://patents.google.com/patent/US20230054924A1
https://patents.google.com/patent/US20230058448A1

**Reply by crtag** [reply]
I'm not sure this is relevant. The propsal is based on the w3c recommended standard for DID. Or am I missing something?

**Reply by toshiyotmoto** [reply]
It seems trivial, but certain CEO's are making claims https://prnt.sc/JwKVM5zPy24U

If specific influential individuals are making false allegations, then the Foundation's legal counsel needs to address the concerns and matters related to Tort Law expeditiously.

**Reply by crtag** [reply]
Well, due diligence has to be made at the time of architectural/design decisions, but I highly doubt patents in space of standards can be enacted. 
https://www.w3.org/Consortium/Legal/IPR-FAQ-20000620

**Reply by rlp18pharmd** [reply]
I'm not a copyright/patent attorney by any stretch of the imagination but the DiD design and configuration would be specific to that of the XRPL right? If I'm correct, the Mettalicus DiD link you provided above references a completely different blockchain architecture from that of the XRPL. I'm just trying to wrap my head around how an XRPL DiD protocol would infringe upon the copyrights of a completely different network/blockchain DiD protocol/platform. Especially, if the underlying blockchains are different and the DiD configuration/design requirements would be different as well.

**Reply by toshiyotmoto** [reply]
Consider individual components such as computers, routers, switches, and cables. The components themselves are not patentable (novel); however, imagine that a network engineer develops a unique and innovative network architecture that optimizes data transmission between the components in a new way. This could involve a novel routing algorithm, an inventive method for load balancing, or an advanced security protocol that enhances the overall performance, reliability, and security of the computer network, which is patentable (given no prior art exists).

In this scenario, the overall system or architecture – the specific arrangement, interaction, and communication between the non-patentable components – could potentially be patentable, as it represents a novel and inventive solution in the field of computer networks. The patent would cover the unique design and configuration of the network system, rather than the individual components themselves.

In said US17/190,320 patent application, look at the figures and you will see very generic language used (blockchain, api, user profile, etc) which is a broad scope. Patent attorneys rarely pigeon hole themselves into a scope that is too limited. I could go through all the nuances of patentability, but its quite complex. I am NOT implying there are any infringements with any of the patents or applications. All I'm saying is everyone is chasing the same rabbit, and its smart business to have a patent attorney review any architecture to ensure there is no potential liabilities before deploying. Now, if any patent or patent application contains concepts, architectures, workflows that predate when the patent was applied for, then that is prior art and could invalidate a patent or prevent it from being granted - which is the whole purpose of the patent pending stage.

It is highly probable all these patents and applications employ W3C standards; however, it's the unique combination, system, workflow, or architecture that could be patentable.

**Reply by Silkjaer** [reply]
Just to clarify, this proposal is put forward by Ripple, and is yet only a standards draft, and not implemented nor live on any networks. But with concerns like these, the contributor need to make sure it does not infringe any patents before an implementation can be merged into code.

**Comment by toshiyotmoto** [comment]
Has anyone read CanDID: Can-Do Decentralized Identity with Legacy Compatibility, Sybil-Resistance,
and Accountability https://eprint.iacr.org/2020/934.pdf

**Comment by shortthefomo** [comment]
The URI field much like the URI field on NFT's today should be a required field.

**Reply by shortthefomo** [reply]
Can any explanation around why the URI be empty be provided?

**Reply by mvadari** [reply]
In the current version of the spec, you must either have the URI field or the Data field.

**Comment by FlorianBouron** [comment]
Awesome job!

**Comment by shortthefomo** [comment]
The other issue here is this suffers the same lookup issue as to who the owner is like the current NFT standard. Requiring extra tooling for indexes that live off the XRPL. It would be desirable to look up the owner of the DiD via the ledger it's self proving the DiD identifier.

**Reply by mvadari** [reply]
This should no longer be an issue with the current version of the spec.

**Comment by tequdev** [comment]
How did the specification regarding Account Roles go away?

**Comment by ckeshava** [comment]
Hello, I have a few questions.
1. Key-rotation: Individuals/organizations would like to periodically refresh their private/public keys for better security. Can we accomplish that with this DID proposal? Would that be useful?
I know that this can be accomplished with special cryptographic primitives or a new transaction type that updates the publicKey associated with a DID.

On a broader note in the XRPL architecture, I know that ephemeral and masterPublic keys are used in the manifests. Are users provided any way to rotate their keys for account management? 

2. Multiple identities:
If an account needs to hold multiple identities, how would one go about doing that? Do we have any equivalent for "also-known-as" field? (https://www.w3.org/TR/did-core/#also-known-as)

The DIDID appears to be tightly bound to the Account ID, hence it appears that multiple identities is not feasible for a single account.

**Reply by aanchal4** [reply]
1. I'm unsure about the specific question you have regarding key rotation. Is your Q related to the rotation of the master public key of the ledger? Nevertheless, I'll make an attempt to provide an explanation. Within the XRPL context, it's feasible to disable the master key, but rotation is not a supported feature. However, the regular key can be rotated periodically. The DID URI is a function of AccountID or master public key, serving the purpose of ensuring resolution through a blockchain. Implementing DID support does not bring about any alterations to the existing account management procedures on the XRPL platform. It's worth noting that DIDs can be linked to multiple keys (resolvable through DID URI). Given that modifications can be made to the public keys associated with the DID, changing the regular key will also lead to a corresponding adjustment in the DID reference.

2. DID is unique for an Account. DID is a URI that can be resolved to a DID document. DID document may contain as many "also-known-as" fields to associate multiple URIs to the master DID.

**Reply by ckeshava** [reply]
yes, my question was about rotation of master public keys. Thanks for your answers.

**Reply by intelliot** [reply]
There is no need to rotate the master key because it has no important privileges over a regular key or a signer list. Key rotation is fully supported by the XRPL already, by assigning a [regular key](https://xrpl.org/assign-a-regular-key-pair.html) (or signer list) and then disabling the [master key](https://xrpl.org/disable-master-key-pair.html).

**Comment by ckeshava** [comment]
3. Right-to-forget: As per several privacy laws (GDPR, HIPAA, etc) across the globe, users should be able to permanently delete the data from a system. Hypothetically, if a user places "Personal Identifiable Information" in the data field of the DID object and needs to delete it at a later point of time, is that possible? Even if the DID object is deleted, such data can always be viewed in the previous ledgers.
(I know that this is a general critique of blockchains, given that they maintain provenance)

**Reply by shortthefomo** [reply]
The DiD is simply a URI it does not save anything. The referenced URI would have the data. It would be up to the implementation(s) of the interface where this data is stored and how/if can be removed.

**Reply by ckeshava** [reply]
Yes, I understand that. I'm talking about the `Data` field, which is 256 bytes in length. If that field is used to hold the `Subject` or any sensitive data, it is stored on all the validators forever. The `Data` field is un-deletable.

**Reply by mvadari** [reply]
The `Data` field is not un-deleteable, that field can also be deleted.

**Reply by aanchal4** [reply]
Data field is under the control of the account/DID & it is public knowledge that blockchain data is immutable. It is not supposed to contain PII. It is supposed to contain publicly accessible attestations.  While the ledger cannot prevent individuals from including their PII, this isn't an unprecedented issue. Similar to how any PII could be included in the memo field and remain permanent.

**Reply by ckeshava** [reply]
hmm, I see

**Comment by ckeshava** [comment]
Yes, but it would preserve the Data field in all the previous ledgers, I'm
concerned that goes against the spirit of privacy laws.
Regards,
Keshava

**Comment by adrianhopebailie** [comment]
Hey @aanchal4 this is excellent, thanks for putting it together.

I started playing around with designing a DID method for XRPL a few years ago but limited my design to not requiring any new ledger objects. This is much cleaner and I think could be incredibly powerful.

I think it would be really useful to document some use cases for this where the account owner is an individual but also where the account owner is a business.

E.g. Consider the case of key rotation raised in an earlier comment. A business could publish a DID Document that contains all their public keys used for attestation, authorization etc and then bind that to their account on XRPL using a key that is mostly kept in cold storage. The private signing and auth keys can be stored in an HSM or similar and used for day-to-day operations like signing API requests or issuing verifiable claims.

Another interesting thought is considering this as a way for individuals to effectively be their own Certificate Authority where their master key on XRPL is the root of their own PKI.

**Comment by intelliot** [comment]
note: as of today, this spec needs some updates

**Comment by aanchal4** [comment]
Refer to PR https://github.com/XRPLF/XRPL-Standards/pull/136

**Comment by ThierryThevenet** [comment]
@aanchal4  Hello, any update about the possibility to have a driver for the universal resolver ?
We plan to use xrp DID method for VC issuance (for jwt_vc_json) with OIDC4VC and our wallets https://altme.io

**Reply by mvadari** [reply]
I have a WIP here: https://github.com/mvadari/xrpl-did-resolver
Far from complete, let me know if you have any feedback 🙂
